esp-hal/src/i2c.rs

```diff
@@ -518,65 +520,89 @@ where
         scl.pull_direction(Pull::Up, crate::private::Internal);
 
-        scl.connect_input_to_peripheral(
-            i2c.peripheral.scl_input_signal(),
-            crate::private::Internal,
-        );
-        scl.connect_peripheral_to_output(
-            i2c.peripheral.scl_output_signal(),
-            crate::private::Internal,
-        );
+        scl.connect_input_to_peripheral(i2c.i2c.scl_input_signal(), crate::private::Internal);
+        scl.connect_peripheral_to_output(i2c.i2c.scl_output_signal(), crate::private::Internal);
 
         sda.set_to_open_drain_output(crate::private::Internal);
         sda.enable_input(true, crate::private::Internal);
         sda.pull_direction(Pull::Up, crate::private::Internal);
 
-        sda.connect_input_to_peripheral(
-            i2c.peripheral.sda_input_signal(),
-            crate::private::Internal,
-        );
-        sda.connect_peripheral_to_output(
-            i2c.peripheral.sda_output_signal(),
-            crate::private::Internal,
-        );
+        sda.connect_input_to_peripheral(i2c.i2c.sda_input_signal(), crate::private::Internal);
+        sda.connect_peripheral_to_output(i2c.i2c.sda_output_signal(), crate::private::Internal);
 
-        i2c.peripheral.setup(frequency, timeout);
+        i2c.i2c.setup(frequency, timeout);
         i2c
     }
 
     fn internal_set_interrupt_handler(&mut self, handler: InterruptHandler) {
-        unsafe { crate::interrupt::bind_interrupt(T::interrupt(), handler.handler()) };
-        unwrap!(crate::interrupt::enable(T::interrupt(), handler.priority()));
+        unsafe { crate::interrupt::bind_interrupt(self.i2c.interrupt(), handler.handler()) };
+        unwrap!(crate::interrupt::enable(
+            self.i2c.interrupt(),
+            handler.priority()
+        ));
     }
 
     fn internal_recover(&self) {
-        PeripheralClockControl::reset(T::peripheral());
-        PeripheralClockControl::enable(T::peripheral());
+        PeripheralClockControl::reset(self.i2c.peripheral());
+        PeripheralClockControl::enable(self.i2c.peripheral());
+
+        self.i2c.setup(self.frequency, self.timeout);
+    }
+}
+
+impl<'d> I2c<'d, Blocking> {
+    /// Create a new I2C instance
+    /// This will enable the peripheral but the peripheral won't get
+    /// automatically disabled when this gets dropped.
+    pub fn new<SDA: PeripheralOutput + PeripheralInput, SCL: PeripheralOutput + PeripheralInput>(
+        i2c: impl Peripheral<P = impl Into<AnyI2c>> + 'd,
+        sda: impl Peripheral<P = SDA> + 'd,
+        scl: impl Peripheral<P = SCL> + 'd,
+        frequency: HertzU32,
+    ) -> Self {
+        Self::new_with_timeout(i2c, sda, scl, frequency, None)
+    }
 
-        self.peripheral.setup(self.frequency, self.timeout);
+    /// Create a new I2C instance with a custom timeout value.
+    /// This will enable the peripheral but the peripheral won't get
+    /// automatically disabled when this gets dropped.
+    pub fn new_with_timeout<
```

Comment:
This constructor is terrible and in the current implementation it's pretty easy to introduce a `.with_timeout()` instead. Anybody knows why this wasn't done initially, e.g. was the timeout setting more intertwined in the past with `configure_clock`'s contents?

Reply:
Yeah I guess the half_cycle calculation makes sense and it's needed for the timeout. Although I don't quite understand what happens here - why do we set timeout_enable = true even if timeout is None?